var/spack/repos/builtin/packages/petsc/package.py

```diff
@@ -576,54 +578,118 @@ def headers(self):
+                   '-snes_monitor']
+        purpose_str = 'test: run ex3k.kokkos example with +kokkos'
+
+        self.run_test(runexe,
+                      options=runopt + testexe,
+                      purpose=purpose_str,
+                      work_dir=w_dir)
 
     def test(self):
         # solve Poisson equation in 2D to make sure nothing is broken:
-        spec = self.spec
         env['PETSC_DIR'] = self.prefix
         env['PETSC_ARCH'] = ''
-        if ('+mpi' in spec):
-            runexe = Executable(join_path(spec['mpi'].prefix.bin,
+        if ('+mpi' in self.spec):
+            runexe = Executable(join_path(self.spec['mpi'].prefix.bin,
                                           'mpiexec')).command
-            runopt = ['-n', '4']
         else:
-            runexe = Executable(join_path(self.prefix,
-                                          'lib/petsc/bin/petsc-mpiexec.uni')).command
-            runopt = ['-n', '1']
-        w_dir = join_path(self.install_test_root, 'src/ksp/ksp/tutorials')
-        with working_dir(w_dir):
-            testexe = ['ex50', '-da_grid_x', '4', '-da_grid_y', '4']
-            testdict = {
-                None: [],
-                '+superlu-dist':
-                ['-pc_type', 'lu', '-pc_factor_mat_solver_type', 'superlu_dist'],
-                '+mumps':
-                ['-pc_type', 'lu', '-pc_factor_mat_solver_type', 'mumps'],
-                '+hypre':
-                ['-pc_type', 'hypre', '-pc_hypre_type', 'boomeramg'],
-                '+mkl-pardiso':
-                ['-pc_type', 'lu', '-pc_factor_mat_solver_type', 'mkl_pardiso'],
-            }
-            make('ex50', parallel=False)
-            for feature, featureopt in testdict.items():
-                if not feature or feature in spec:
-                    self.run_test(runexe, runopt + testexe + featureopt)
-            if '+cuda' in spec:
-                make('ex7', parallel=False)
-                testexe = ['ex7', '-mat_type', 'aijcusparse',
-                           '-sub_pc_factor_mat_solver_type', 'cusparse',
-                           '-sub_ksp_type', 'preonly', '-sub_pc_type', 'ilu',
-                           '-use_gpu_aware_mpi', '0']
-                self.run_test(runexe, runopt + testexe)
-            make('clean', parallel=False)
-        w_dir = join_path(self.install_test_root, 'src/snes/tutorials')
-        with working_dir(w_dir):
-            if '+kokkos' in spec:
-                make('ex3k', parallel=False)
-                testexe = ['ex3k', '-view_initial', '-dm_vec_type', 'kokkos',
-                           '-dm_mat_type', 'aijkokkos', '-use_gpu_aware_mpi', '0',
-                           '-snes_monitor']
-                self.run_test(runexe, runopt + testexe)
-            make('clean', parallel=False)
+            runexe = Executable(join_path(self.prefix.lib.petsc.bin,
+                                          'petsc-mpiexec.uni')).command
```

Comment:
I suggest making sure the command `runexe` actually exists.

I had trouble building the package but, in the one case I could, the execution part of the test failed with `Failed to find executable /path/to/mpiexec`.  I also did not see `petsc-mpiexec.uni` under `self.prefix.lib.petsc.bin`.  I *did* see `petsc-mpiexec` there though.  If that is usable even if `+mpi`, then you might want to have a list of possible `mpiexec`s and run with the first one that actually exists (otherwise fail).

Reply:
Interesting.  Looks like using `petscmpiexec` is not an option.  When I tried I get 

```
  >> 59    Error build location not found! Please set PETSC_DIR and PETSC_ARCH 
           correctly for this build.
```

@balay Is the above error tied to my attempt to use `petscmpiexec` when the `mpi` I used to build the software doesn't have an `mpiexec`?

Reply:
Or do these environment variables always need to be set if using a `petsc`-provided `mpiexec`?

Reply:
Sorry - I'm not in sync with all changes in this PR

> when the mpi I used to build the software doesn't have an mpiexec

This is more of a spack issue. How does one run MPI tests when there is no mpiexec? Some pkgs assume srun - and have hardcoded this in their test section. Perhaps we should be able to query spack to get this info?

And `petscmpiexec` is just a convenience wrapper over the primary `mpiexec` [from the MPI  PETSc is built with] - and its not a subtitute when primary mpiexe is absent

>  I also did not see petsc-mpiexec.uni under self.prefix.lib.petsc.bin

this gets installed only for `petsc~mpi` build. And one doesn't need that - its there to provide a regular mpiexec interface to sequential runs

>   Error build location not found! Please set PETSC_DIR and PETSC_ARCH correctly for this build.

Hm - this should be set by: [for ex: as in develop branch]
```
    def test(self):
        # solve Poisson equation in 2D to make sure nothing is broken:                                                                                                                                             
        spec = self.spec
        env['PETSC_DIR'] = self.prefix
        env['PETSC_ARCH'] = ''
```